File: pkg/config/constraints.go
package config

import (
	"fmt"

	"gopkg.in/yaml.v3"
)

// ValidateConstraints validates configuration against schema constraints
func ValidateConstraints(cfg Config) []string {
	var errors []string

	// Load constraints from schema
	constraints, err := loadSchemaConstraints()
	if err != nil {
		return []string{fmt.Sprintf("Failed to load constraints: %v", err)}
	}

	// Check each constraint
	for _, constraint := range constraints {
		// Mutually exclusive fields - only check if at least one field is present
		if len(constraint.MutuallyExclusive) >= 2 {
			if anyFieldPresent(cfg, constraint.MutuallyExclusive) {
				if err := checkMutuallyExclusive(cfg, constraint.MutuallyExclusive); err != nil {
					errors = append(errors, err.Error())
				}
			}
		}

		// One-of constraints - always check (must have exactly one)
		if len(constraint.OneOf) > 0 {
			if err := checkOneOfFields(cfg, constraint.OneOf, constraint.Error); err != nil {
				errors = append(errors, err.Error())
			}
		}
	}

	return errors
}

// anyFieldPresent checks if any of the fields exist in config
func anyFieldPresent(cfg Config, fields []string) bool {
	for _, field := range fields {
		val, exists := cfg[field]
		if exists && val != nil && val != "" {
			return true
		}
	}
	return false
}

// checkMutuallyExclusive verifies only one of the fields is set
func checkMutuallyExclusive(cfg Config, fields []string) error {
	setFields := []string{}

	for _, field := range fields {
		val, exists := cfg[field]
		if exists && val != nil && val != "" {
			setFields = append(setFields, field)
		}
	}

	if len(setFields) > 1 {
		return fmt.Errorf("fields %v are mutually exclusive, but %v are set", fields, setFields)
	}

	return nil
}

// checkOneOfFields verifies exactly one of the fields is set
func checkOneOfFields(cfg Config, fields []string, errorMsg string) error {
	setCount := 0
	var setFields []string

	for _, field := range fields {
		if isFieldSet(cfg, field) {
			setCount++
			setFields = append(setFields, field)
		}
	}

	if setCount != 1 {
		if errorMsg != "" {
			return fmt.Errorf("%s", errorMsg)
		}
		if setCount == 0 {
			return fmt.Errorf("exactly one of %v must be set, but none are set", fields)
		}
		return fmt.Errorf("exactly one of %v must be set, but %v are set", fields, setFields)
	}

	return nil
}

// ConstraintDef represents a constraint from the schema
type ConstraintDef struct {
	MutuallyExclusive []string `yaml:"mutually_exclusive" json:"mutually_exclusive,omitempty"`
	OneOf             []string `yaml:"one_of" json:"one_of,omitempty"`
	Error             string   `yaml:"error" json:"error,omitempty"`
}

// Alias for backward compatibility
type constraintDef = ConstraintDef

// isFieldSet checks if a field has a truthy value
func isFieldSet(cfg Config, field string) bool {
	val, exists := cfg[field]
	if !exists || val == nil {
		return false
	}

	// Check boolean fields
	if boolVal, ok := val.(bool); ok {
		return boolVal
	}

	// Check string fields
	if strVal, ok := val.(string); ok {
		return strVal != ""
	}

	return true
}

// schemaConstraints represents the constraints section
type schemaConstraints struct {
	Constraints []constraintDef `yaml:"constraints"`
}

var cachedConstraints []constraintDef

// LoadConstraints loads constraints from the schema file (exported for API)
func LoadConstraints() ([]ConstraintDef, error) {
	return loadSchemaConstraints()
}

// loadSchemaConstraints loads constraints from the schema file
func loadSchemaConstraints() ([]constraintDef, error) {
	if cachedConstraints != nil {
		return cachedConstraints, nil
	}

	// Use the embedded schema data from schema_loader.go
	var schema schemaConstraints
	if err := yaml.Unmarshal(schemaData, &schema); err != nil {
		return nil, fmt.Errorf("failed to parse schema: %w", err)
	}

	cachedConstraints = schema.Constraints
	return cachedConstraints, nil
}

File: cmd/web/static/js/constraints.js
// constraints.js - Client-side constraint validation

let cachedConstraints = null;

// Load constraints from schema
async function loadConstraints() {
    if (cachedConstraints) {
        return cachedConstraints;
    }

    try {
        const response = await fetch('/api/schema');
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        const data = await response.json();
        cachedConstraints = data.constraints || [];
        return cachedConstraints;
    } catch (error) {
        console.error('Failed to load constraints:', error);
        return [];
    }
}

// Check if any of the fields exist in config
function anyFieldPresent(config, fields) {
    for (const field of fields) {
        const val = config[field];
        if (val !== undefined && val !== null && val !== '') {
            return true;
        }
    }
    return false;
}

// Extract field names from condition strings
// Example: "NO_DISKS_FOR_CLUSTER == true && CLUSTER_DISKS == ''" -> ["NO_DISKS_FOR_CLUSTER", "CLUSTER_DISKS"]
function extractFieldsFromConditions(conditions) {
    const fieldMap = new Map();

    for (const condition of conditions) {
        // Split by && for AND logic
        const parts = condition.split(' && ');

        for (let part of parts) {
            part = part.trim();

            // Extract field name (everything before == or !=)
            if (part.includes(' == ')) {
                const tokens = part.split(' == ');
                if (tokens.length === 2) {
                    fieldMap.set(tokens[0].trim(), true);
                }
            } else if (part.includes(' != ')) {
                const tokens = part.split(' != ');
                if (tokens.length === 2) {
                    fieldMap.set(tokens[0].trim(), true);
                }
            }
        }
    }

    return Array.from(fieldMap.keys());
}

// Get config value as normalized string
function getConfigValue(config, key) {
    const val = config[key];
    if (val === undefined || val === null) {
        return '';
    }

    // Convert to string representation
    if (typeof val === 'boolean') {
        return val ? 'true' : 'false';
    }
    if (typeof val === 'string') {
        return val;
    }
    return String(val);
}

// Evaluate a boolean condition string
// Format: "KEY == value && KEY2 == value2" or "KEY != value"
function evaluateCondition(condition, config) {
    // Split by && for AND logic
    const parts = condition.split(' && ');

    for (let part of parts) {
        part = part.trim();

        // Check for != operator
        if (part.includes(' != ')) {
            const tokens = part.split(' != ');
            if (tokens.length !== 2) {
                return false;
            }

            const key = tokens[0].trim();
            let expectedValue = tokens[1].trim();
            expectedValue = expectedValue.replace(/^["']|["']$/g, ''); // Remove quotes

            const actualValue = getConfigValue(config, key);
            if (actualValue === expectedValue) {
                return false; // Values are equal, so != is false
            }
            continue;
        }

        // Check for == operator
        if (part.includes(' == ')) {
            const tokens = part.split(' == ');
            if (tokens.length !== 2) {
                return false;
            }

            const key = tokens[0].trim();
            let expectedValue = tokens[1].trim();
            expectedValue = expectedValue.replace(/^["']|["']$/g, ''); // Remove quotes

            const actualValue = getConfigValue(config, key);
            if (actualValue !== expectedValue) {
                return false;
            }
            continue;
        }

        return false;
    }

    return true;
}

// Check mutually exclusive constraint
function checkMutuallyExclusive(config, fields) {
    const setFields = [];

    for (const field of fields) {
        const val = config[field];
        if (val !== undefined && val !== null && val !== '') {
            setFields.push(field);
        }
    }

    if (setFields.length > 1) {
        return `Fields ${fields.join(', ')} are mutually exclusive, but ${setFields.join(', ')} are set`;
    }

    return null;
}

// Check if a field is set (has a truthy value)
function isFieldSet(config, field) {
    const val = config[field];
    if (val === undefined || val === null) {
        return false;
    }

    // Check boolean fields
    if (typeof val === 'boolean') {
        return val;
    }

    // Check string fields
    if (typeof val === 'string') {
        return val !== '';
    }

    return true;
}

// Check one-of constraint for fields
function checkOneOfFields(config, fields, errorMsg) {
    let setCount = 0;
    const setFields = [];

    for (const field of fields) {
        if (isFieldSet(config, field)) {
            setCount++;
            setFields.push(field);
        }
    }

    if (setCount !== 1) {
        if (errorMsg) {
            return errorMsg;
        }
        if (setCount === 0) {
            return `Exactly one of ${fields.join(', ')} must be set, but none are set`;
        }
        return `Exactly one of ${fields.join(', ')} must be set, but ${setFields.join(', ')} are set`;
    }

    return null;
}

// Validate all constraints
async function validateConstraints(config) {
    const errors = [];
    const constraints = await loadConstraints();

    for (const constraint of constraints) {
        // Mutually exclusive fields - only check if at least one field is present
        if (constraint.mutually_exclusive && constraint.mutually_exclusive.length >= 2) {
            if (anyFieldPresent(config, constraint.mutually_exclusive)) {
                const error = checkMutuallyExclusive(config, constraint.mutually_exclusive);
                if (error) {
                    errors.push(error);
                }
            }
        }

        // One-of constraints - only check if any relevant field is present
        if (constraint.one_of && constraint.one_of.length > 0) {
            if (anyFieldPresent(config, constraint.one_of)) {
                const error = checkOneOfFields(config, constraint.one_of, constraint.error);
                if (error) {
                    errors.push(error);
                }
            }
        }
    }

    return errors;
}
